### DISCUSSION TITLE
2.2.1版本，数据插入速度在200条/s，速度还能再提升吗？

### DISCUSSION THREAD

**Discussion by ingale726** [milvus-io/milvus#21361] [Q&A and General discussion] [answered]
大家都是怎么插入数据的?
我是这么插入的数据的
```
  entities = [
      [line],  # field random
      [embedding],  # field embeddings
  ]
  result = hello_milvus.insert(entities)
```
你们的几千条/s都是怎么实现的？

**Comment by bingbb47** [comment]
entities = [
      [line,line,line,...],  # field random
      [embedding,embedding,embedding,...],  # field embeddings
  ]
  result = hello_milvus.insert(entities)

**Reply by xiaofan-luan** [reply]
in

> 大家都是怎么插入数据的? 我是这么插入的数据的
> 
> ```
>   entities = [
>       [line],  # field random
>       [embedding],  # field embeddings
>   ]
>   result = hello_milvus.insert(entities)
> ```
> 
> 你们的几千条/s都是怎么实现的？

insert by batch or concurrently can achieve

**Comment by ingale726** [comment]
Are there rules and recommendations on the length of the list?
like this:
[embedding,embedding,embedding,...]
the length of the list is 256 or 512 ,is ok?

**Comment by bingbb47** [comment, marked as answer]
i think the batchsize depends on the limitation of grpc
This is my example

```
ids=0
batchsize = 10000
dim = 128
max_abs_scaler = preprocessing.MaxAbsScaler()
for i in tqdm(range(batchnum)):
    ts = time.time()
    idbatch = [ ids+i for i in range(batchsize)]
    scalarbatch = [random.randint(1, 100) for _ in range(batchsize)]
    vectorbatch = []
    for bi in range(batchsize):
        v = np.random.rand(dim)*np.random.choice([1,-1])
        normalized_v = v/np.linalg.norm(v)
        vectorbatch.append(normalized_v)
    data = [idbatch, scalarbatch, vectorbatch]
    collection.insert(data)
    te = time.time()
    print(f"time {te-ts}s")
    ids += batchsize
```

**Comment by ingale726** [comment]
I use the method of batch insertion, and the qps at present is around 400, which can satisfy my use, but is there a faster way? I saw someone say you can go over 1000 qps, how do you do that? Welcome to discuss
